### PR TITLE
Workaround leaflet icon/draggability bug

### PR DESCRIPTION
**Pull request by codyjroberts**
Addresses a bug in leaflet where a marker fails to disable dragging if
it's icon is changed.

https://github.com/miguelcobain/ember-leaflet/issues/75

### CODE REVIEW

File: addon/mixins/icon.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import Ember from 'ember';
+const { observer } = Ember;
+
+export default Ember.Mixin.create({
+
+  iconDidChange: observer('icon', function() {
```
Comment (codyjroberts) [inline addon/mixins/icon.js]: It may make more sense to add the observer in the DraggabilityMixin.

Thoughts?

**Comment by miguelcobain** [comment]
Thanks you for the PR.

Afaik, ember observers don't guarantee order, so it is "pure chance" this works because if this ran *before* the icon changing observer, the problemas would remain.

What we need to do is to customize the icon changing logic.

1. Remove `icon` from `leafletProperties` on marker-layer. This is a marker layer concern. We will no longer rely on the default ember-leaflet observer setup.
2. Create an observer that changes the icon *and* seta the correct draggability.

Let me know if you need more info.

**Comment by codyjroberts** [comment]
Ah, yes that's a major issue!  I wasn't aware of this.  I'll give it a shot.

File: addon/mixins/icon.js (hunk 1 above)
Comment (miguelcobain) [inline addon/mixins/icon.js]: Lets remove Icon mixin and put this into marker-layer directly, ideally with a comment to explain why we're writing this observer manually. This is a marker-layer concern, so no need to be afraid of putting it there. :)

File: tests/integration/components/marker-layer-test.js [hunk 2]
```diff
@@ -110,6 +110,32 @@ test('marker updates dragging', function(assert) {
...
+
+    //pre-conditions
+   assert.ok(marker._layer.dragging.enabled(), 'marker dragging enabled');
+
+   this.set('draggable', false);
+   this.set('currentIconUrl', 'another-custom-url.png');
+
+   assert.equal(marker._layer.options.icon.options.iconUrl, 'another-custom-url.png');
+   assert.equal(marker._layer.dragging.enabled(), false, 'marker dragging disabled');
+});
```
Comment (miguelcobain) [inline tests/integration/components/marker-layer-test.js]: The tests I envisioned would be:

1. test 1 - negative draggability is preserved on icon changes
  1. Start with icon 1 and draggable false
  2. assert that draggability is off (pre-condition)
  3. change to icon 2
  4. assert that draggability is still off
1. test 2 - positive draggability is preserved on icon changes
  1. Start with icon 1 and draggable true
  2. assert that draggability is on (pre-condition)
  3. change to icon 2
  4. assert that draggability is still on

Makes sense?

File: addon/mixins/icon.js (hunk 1 above)
Comment (codyjroberts) [inline addon/mixins/icon.js]: Ah yeah that does make more sense.

File: tests/integration/components/marker-layer-test.js (hunk 2 above)
Comment (codyjroberts) [inline tests/integration/components/marker-layer-test.js]: Yes, I'll add those.  However, both tests you just described pass on master.

Comment (miguelcobain) [inline tests/integration/components/marker-layer-test.js]: 😐  Are you sure? How are you checking the draggability?

Comment (miguelcobain) [inline tests/integration/components/marker-layer-test.js]: Just to make sure your using `enabled()` and not `enable()`: http://leafletjs.com/reference.html#ihandler-enabled

Comment (miguelcobain) [inline tests/integration/components/marker-layer-test.js]: Trying to understand why these tests pass in master...

Comment (codyjroberts) [inline tests/integration/components/marker-layer-test.js]: I based my original test on what I was experiencing in my application.  I was changing the icon to notify a change in draggability.

Comment (miguelcobain) [inline tests/integration/components/marker-layer-test.js]: I may have understood correctly the bug in leaflet. The problem seems that it retains the original draggability option.

**Comment by miguelcobain** [comment]
I've figured it out. The problem only happens on subsequent draggability changes, not when instantiating the marker.

Please replace all tests by this one:
```js
// Leaflet bug. More info: https://github.com/Leaflet/Leaflet/issues/3807
test('marker retains draggability options when icon changes', function(assert) {

  let icon1 = L.divIcon({className: 'my-div-icon-1'});
  let icon2 = L.divIcon({className: 'my-div-icon-2'});

  this.set('markerCenter', locations.nyc);
  this.set('draggable', true);
  this.set('currentIcon', icon1);

  this.render(hbs`
    {{#leaflet-map zoom=zoom center=center}}
      {{marker-layer location=markerCenter draggable=draggable icon=currentIcon}}
    {{/leaflet-map}}
  `);

    //pre-conditions
   assert.equal(marker._layer.dragging.enabled(), true, 'marker dragging enabled');

   this.set('draggable', false);
   assert.equal(marker._layer.dragging.enabled(), false, 'marker dragging disabled');

   this.set('currentIcon', icon2);
   assert.equal(marker._layer.dragging.enabled(), false, 'marker dragging is still disabled');
});
```

File: addon/components/marker-layer.js [hunk 3]
```diff
@@ -2,6 +2,7 @@ import BaseLayer from 'ember-leaflet/components/base-layer';
 import DraggabilityMixin from 'ember-leaflet/mixins/draggability';
 import PopupMixin from 'ember-leaflet/mixins/popup';
 import toLatLng from 'ember-leaflet/macros/to-lat-lng';
+import observer from 'ember-metal/observer';
```
Comment (miguelcobain) [inline addon/components/marker-layer.js]: Let's use `const { observer } = Ember;` like we do on the rest of the project, please.
We'll update everything when the modules story is settled for good.

**Comment by miguelcobain** [comment]
@codyjroberts only that little issue, and we're good to go!

**Comment by codyjroberts** [comment]
Excellent fixing that now.

**Comment by miguelcobain** [comment]
@codyjroberts Many thanks for this PR!

I've just published `ember-leaflet@2.2.7` with this fix. :shipit:

**Pull request merged**